File: src/web_bak/src/api/calcReportCategory.ts
import { httpClient } from 'src/api/base/httpClient'
import type { ICategoryInfo } from 'src/components/categoryList/types'

/**
 * 获取所有计算报告分类
 * @returns 分类列表
 */
export function getCalcReportCategories() {
  return httpClient.get<ICategoryInfo[]>('/calc-report-category/list')
}

/**
 * 获取单个计算报告分类
 * @param categoryOid 分类 OID
 * @returns 分类信息
 */
export function getCalcReportCategory(categoryOid: string) {
  return httpClient.get<ICategoryInfo>(`/calc-report-category/${categoryOid}`)
}

/**
 * 创建计算报告分类
 * @param data 分类信息
 * @returns 创建后的分类信息
 */
export function createCalcReportCategory(data: {
  name: string
  order?: number
  description?: string | null
  cover?: string | null
}) {
  return httpClient.post<ICategoryInfo>('/calc-report-category', {
    data
  })
}

/**
 * 更新计算报告分类
 * @param categoryOid 分类ID
 * @param data 分类信息
 * @returns 更新后的分类信息
 */
export function updateCalcReportCategory(categoryOid: string, data: Partial<ICategoryInfo>) {
  return httpClient.put<ICategoryInfo>(`/calc-report-category/${categoryOid}`, {
    data
  })
}

/**
 * 删除计算报告分类
 * @param categoryOid 分类ID
 */
export function deleteCalcReportCategory(categoryOid: string) {
  return httpClient.delete(`/calc-report-category/${categoryOid}`)
}

/**
 * 获取或创建默认分类
 * 如果用户没有任何有效分类，则使用默认名称创建一个分类后返回
 * @param defaultCategoryName 默认分类名称
 * @returns 分类信息
 */
export function getOrCreateDefaultCategory(defaultCategoryName: string) {
  return httpClient.post<ICategoryInfo>('/calc-report-category/default', {
    data: { defaultCategoryName }
  })
}

File: src/api/app/controller/calc/calc_report_category.py
"""HTTP endpoints for calculation-report categories."""

from fastapi import APIRouter, Depends
from sqlalchemy.ext.asyncio import AsyncSession

from app.controller.calc.calc_report_dto import (
    CalcReportCategoryReqDTO,
    CalcReportCategoryResDTO,
    CategoryOrderDTO,
    CategoryStateDTO,
)
from app.controller.depends import get_session, get_token_payload
from app.response.response_result import ResponseResult, ok
from app.service import calc_report_category_service
from utils.jwt_helper import TokenPayloads

router = APIRouter(prefix="/v1/calc-report-category", tags=["calc-report-category"])


@router.get("")
async def list_calc_report_categories(
    tokenPayloads: TokenPayloads = Depends(get_token_payload),
    session: AsyncSession = Depends(get_session),
) -> ResponseResult[list[CalcReportCategoryResDTO]]:
    """List current user's active report categories."""
    return ok(
        data=await calc_report_category_service.list_categories(
            tokenPayloads.id, session
        )
    )


@router.post("")
async def create_calc_report_category(
    request: CalcReportCategoryReqDTO,
    tokenPayloads: TokenPayloads = Depends(get_token_payload),
    session: AsyncSession = Depends(get_session),
) -> ResponseResult[CalcReportCategoryResDTO]:
    """Create a report category."""
    return ok(
        data=await calc_report_category_service.create_category(
            tokenPayloads.id, request, session
        )
    )


@router.post("/default")
async def get_default_calc_report_category(
    request: CalcReportCategoryReqDTO,
    tokenPayloads: TokenPayloads = Depends(get_token_payload),
    session: AsyncSession = Depends(get_session),
) -> ResponseResult[CalcReportCategoryResDTO]:
    """Get or create the named default category."""
    return ok(
        data=await calc_report_category_service.get_or_create_default_category(
            tokenPayloads.id, request.name, session
        )
    )


@router.put("/order")
async def reorder_calc_report_categories(
    orders: list[CategoryOrderDTO],
    tokenPayloads: TokenPayloads = Depends(get_token_payload),
    session: AsyncSession = Depends(get_session),
) -> ResponseResult[list[CalcReportCategoryResDTO]]:
    """Replace explicit category sort orders."""
    return ok(
        data=await calc_report_category_service.reorder_categories(
            tokenPayloads.id, orders, session
        )
    )


@router.put("/{categoryOid}/state")
async def update_calc_report_category_state(
    categoryOid: str,
    request: CategoryStateDTO,
    tokenPayloads: TokenPayloads = Depends(get_token_payload),
    session: AsyncSession = Depends(get_session),
) -> ResponseResult[CalcReportCategoryResDTO]:
    """Update category pinning and visibility state."""
    return ok(
        data=await calc_report_category_service.update_category_state(
            tokenPayloads.id, categoryOid, request, session
        )
    )


@router.post("/{categoryOid}/access")
async def record_calc_report_category_access(
    categoryOid: str,
    tokenPayloads: TokenPayloads = Depends(get_token_payload),
    session: AsyncSession = Depends(get_session),
) -> ResponseResult[CalcReportCategoryResDTO]:
    """Record one category activation for LFU-Aging ordering."""
    return ok(
        data=await calc_report_category_service.record_category_access(
            tokenPayloads.id, categoryOid, session
        )
    )


@router.put("/{categoryOid}")
async def update_calc_report_category(
    categoryOid: str,
    request: CalcReportCategoryReqDTO,
    tokenPayloads: TokenPayloads = Depends(get_token_payload),
    session: AsyncSession = Depends(get_session),
) -> ResponseResult[CalcReportCategoryResDTO]:
    """Update category metadata."""
    return ok(
        data=await calc_report_category_service.update_category(
            tokenPayloads.id, categoryOid, request, session
        )
    )


@router.delete("/{categoryOid}")
async def delete_calc_report_category(
    categoryOid: str,
    tokenPayloads: TokenPayloads = Depends(get_token_payload),
    session: AsyncSession = Depends(get_session),
) -> ResponseResult[None]:
    """Soft-delete an empty category."""
    await calc_report_category_service.delete_category(
        tokenPayloads.id, categoryOid, session
    )
    return ok()
